algosdk/source_map.py

```diff
@@ -0,0 +1,101 @@
+from typing import Dict, Any, List, Tuple
+
+from algosdk.error import SourceMapVersionError
+
+
+class SourceMap:
+    """
+    Decodes a VLQ-encoded source mapping between PC values and TEAL source code lines.
+    Spec available here: https://sourcemaps.info/spec.html
+
+    Args:
+        source_map (dict(str, Any)): source map JSON from algod
+    """
+
+    def __init__(self, source_map: Dict[str, Any]):
+
+        self.version: int = source_map["version"]
+
+        if self.version != 3:
+            raise SourceMapVersionError(self.version)
+
+        self.sources: List[str] = source_map["sources"]
+
+        if "mapping" in source_map:
+            # For backwards compat until go-algorand has the change then we should remove
+            self.mappings: str = source_map["mapping"]
+        else:
+            self.mappings: str = source_map["mappings"]
+
+        pc_list = [
+            _decode_int_value(raw_val) for raw_val in self.mappings.split(";")
+        ]
+
+        # Initialize with 0,0 for pc/line
+        self.pc_to_line: Dict[int, int] = {0: 0}
+        self.line_to_pc: Dict[int, List[int]] = {0: [0]}
+
+        last_line = 0
+        for index, line_num in enumerate(pc_list):
+            if line_num is not None:  # be careful for '0' checks!
+                if line_num not in self.line_to_pc:
+                    self.line_to_pc[line_num] = []
+                self.line_to_pc[line_num].append(index)
+                last_line = line_num
+
+            self.pc_to_line[index] = last_line
+
+    def get_line_for_pc(self, pc: int) -> int:
+        return self.pc_to_line[pc]
```

Comment:
@barnjamin What's the desired behavior when the key (PC) doesn't exist?

As written, `get_line_for_pc` and `get_pcs_for_line` throw `KeyError`.  I'm double checking if the behavior is by-design.

If not, propose switching to `dict.get` and optionally, adding `Optional` return type hint to both methods.

Reply:
Aside:  I think the other SDKs warrant consideration too.  

For example, the as written Go implementation returns the Map's zero value when provided a non-existent key.  I imagine we want to return `Nil`.

Reply:
Additionally, and optionally I'd consider adding a non-Cucumber unit test to enforce the desired behavior.

Reply:
I've no preference on how we handle this so I'll lean towards reviewers opinions

Reply:
I also think we should avoid returning an `Exception`. I would propose `dict.get(pc, None)` and return a `None` or (less preferred) return a deepcopy of the dictionary so users can handle their own errors.

Reply:
I will up with returning `None`, but we may still want to expose `line_to_pc` or `pc_to_line` for someone wants to iterate through the structure.